crates/ruma-client-api/src/discovery/discover_homeserver.rs

```diff
@@ -112,3 +131,163 @@ impl TileServerInfo {
+    pub service_url: String,
+}
+
+/// Information about a custom RTC focus type.
+#[doc(hidden)]
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub struct CustomRtcFocusInfo {
+    /// The type of RTC focus.
+    #[serde(rename = "type")]
+    pub focus_type: String,
+
+    /// Remaining RTC focus data.
+    #[serde(flatten)]
+    pub data: JsonObject,
+}
+
+#[cfg(test)]
+mod tests {
+    #[cfg(feature = "unstable-msc4143")]
+    mod rtc_foci {
+        use serde_json::{from_value as from_json_value, json, to_value as to_json_value};
+
+        use super::super::RtcFocusInfo;
+
+        #[test]
+        fn test_live_kit_rtc_focus_deserialization() {
+            let json = json!({
+                "type": "livekit",
+                "livekit_service_url": "https://livekit.example.com"
+            });
+
+            let focus: RtcFocusInfo = from_json_value(json).unwrap();
+
+            match focus {
+                RtcFocusInfo::LiveKit(info) => {
+                    assert_eq!(info.service_url, "https://livekit.example.com");
+                }
+                RtcFocusInfo::_Custom(_) => panic!("Expected LiveKit focus"),
+            }
+        }
+
+        #[test]
+        fn test_live_kit_rtc_focus_serialization() {
+            let focus = RtcFocusInfo::live_kit("https://livekit.example.com".to_owned());
+
+            let json = to_json_value(&focus).unwrap();
+
+            assert_eq!(
+                json,
+                json!({
+                    "type": "livekit",
+                    "livekit_service_url": "https://livekit.example.com"
+                })
+            );
+        }
+
+        #[test]
+        fn test_custom_rtc_focus_serialization() {
+            let json = json!({
+                "type": "some-focus-type",
+                "additional-type-specific-field": "https://my_focus.domain",
+                "another-additional-type-specific-field": ["with", "Array", "type"]
+            });
+
+            let focus: RtcFocusInfo = from_json_value(json.clone()).unwrap();
+
+            match &focus {
+                RtcFocusInfo::LiveKit(_) => panic!("Expected Custom focus"),
+                RtcFocusInfo::_Custom(info) => {
+                    assert_eq!(info.focus_type, "some-focus-type");
+
+                    let data = &info.data;
+                    assert_eq!(data["additional-type-specific-field"], "https://my_focus.domain");
+
+                    let array_values: Vec<&str> = data["another-additional-type-specific-field"]
+                        .as_array()
+                        .unwrap()
+                        .iter()
+                        .map(|v| v.as_str().unwrap())
+                        .collect();
+                    assert_eq!(array_values, vec!["with", "Array", "type"]);
+
+                    assert!(!data.contains_key("type"));
+                }
+            }
```

Comment:
Here It think that it makes more sense to test the accessors (`focus_type()`, `data()`) instead of matching on the variant, to match what a user of the library should do.